Add balancer unit tests

zrdaley opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

Currently no unit tests exist for ingress-nginx/rootfs/etc/nginx/lua/balancer.lua. This PR adds unit tests for the file.

**Which issue this PR fixes**: N/A

aledbf commented:
/approve

aledbf commented:
@zrdaley thank you for doing this

ElvinEfendi commented on rootfs/etc/nginx/lua/test/balancer_test.lua:
```diff
@@ -1,3 +1,299 @@
+        local ewma_after_balance_spy = spy.on(mock_ewma, "after_balance")
+
+        assert.has_no_errors(balancer.call)
+        assert.spy(backend_get_spy).was_called_with(match.is_table(), "mock_ewma_backend")
```
This does not add any value IMHO and is leaking internal knowledge about implementation.

ElvinEfendi commented on rootfs/etc/nginx/lua/test/balancer_test.lua:
```diff
@@ -1,3 +1,299 @@
+        local ewma_after_balance_spy = spy.on(mock_ewma, "after_balance")
+
+        assert.has_no_errors(balancer.call)
+        assert.spy(backend_get_spy).was_called_with(match.is_table(), "mock_rr_backend")
```
ditto

ElvinEfendi commented on rootfs/etc/nginx/lua/test/balancer_test.lua:
```diff
@@ -1,3 +1,299 @@
+        mock_backends._vals = {}
+
+        local backend_get_spy = spy.on(mock_backends, "get")
+        local ngx_spy = spy.on(_G.ngx, "log")
```
this does not seem to be used anywhere

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer.lua:
```diff
@@ -164,6 +167,9 @@ function _M.call()
 
   local host, port = balance()
+  if not host then
+    return error("balancer did not return a host")
```
hmm it actually does return something - but still when we can not find any upstream to proxy to we should return HTTP 503(ngx.HTTP_SERVICE_UNAVAILABLE)

zrdaley commented:
@ElvinEfendi I addressed all of your concerns.  I was slightly confused about your comments in balancer.lua, so I set ngx.status to HTTP_SERVICE_UNAVAILABLE and also returned an error message.

ElvinEfendi commented on rootfs/etc/nginx/lua/balancer.lua:
```diff
@@ -164,6 +167,10 @@ function _M.call()
   local host, port = balance()
+  if not host then
+    ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+    return error(ngx.HTTP_SERVICE_UNAVAILABLE .. ": service unavailable")
```
just `return` without anything

ElvinEfendi commented:
/lgtm
